Added a test.sql query

rylaix opened the pull request:
SQL discussion

rylaix committed:
Added a test.sql query

fleupold commented on queries/fetch_backruns.sql:
```diff
@@ -1,37 +1,49 @@
+    WHERE to_unixtime(cast(from_unixtime(timestamp / 1000) as timestamp)) > to_unixtime(NOW() - INTERVAL '1' HOUR)
+),
 
+bundles_with_backruns AS (
```
This also fetches bundles without backruns (which is fine, but the name is confusing).

Overall I'm not sure why you need three sub queries here, I'd just go with a single:

```sql
SELECT block_number, timestamp, transactions
FROM mevblocker.raw_bundles t
JOIN ethereum.transactions e
  ON t.blockNumber = e.block_number
  AND from_hex(CAST(json_extract(transactions, '$[0].hash') AS VARCHAR)) = e.hash
  AND block_number >= {{start_block}}
  AND block_number <= {{end_block}}
```

This will fetch all the submitted backruns for all mevblocker transactions on the block they were ultimately included.

rylaix replied:
Are we good with this?:
```SQL
SELECT 
    t.block_number, 
    t.timestamp, 
    t.transactions,
    e.hash AS transaction_hash,
    e."from",
    e.to,
    e.gas_limit,
    e.gas_price,
    e.data AS calldata,
    e.miner,
    e.gas_used,
    e.hash as block_hash,
    e.parent_hash,
    e.base_fee_per_gas
FROM mevblocker.raw_bundles t
JOIN ethereum.transactions e
  ON t.block_number = e.block_number
  AND from_hex(CAST(json_extract(t.transactions, '$[0].hash') AS VARCHAR)) = e.hash
  AND t.block_number >= {{start_block}}
  AND t.block_number <= {{end_block}}
ORDER BY t.timestamp DESC;

```

Only thing is that is kinda hacky, as we must calculate blocks using python (then pass it to dune), and as the script runs periodically (e.g 1 hour or whatever), the amount of blocks we're searching in past for the same amount of time can be different.
Just want to ensure the edge cases and that all meets the requirements.

fleupold replied:
Why do you care about all the fields from `ethereum.transactions`? How is this going to help you in the later parts of the task? 
I think you still need to fetch each block's content from the full node (because you want all the other transactions that were part of the block as well). You could do that in Dune as well, but in any case I would separate the two queries as they are conceptually different things.

Regarding block numbers vs. timestamps I don't have a strong feeling. The goal for this pipeline is to ultimately run for all blocks, so I'd assume it to have some persistence of what was the last processed block and then be able to re-run from there. But of course this is also possible with timestamps.

rylaix replied:
Hm, I think we got misunderstanding here.
I previously were including all possible info to gather whatever it is needed to include all the backruns.
As checked now, your implementation seems better in this point.
In this case, If everything is okay, I will make a new commit with leftover python implementation.
Before that, Do you want to include the query itself on a specific account, or for rn I'll just upload it to mine?

Thanks!

fleupold replied:
You can upload it to yours, as long as it's also checked in, whatever account is running this repo later can re-create the query and specify the query id in the configuraiton.

rylaix replied:
Hello!
Made a commit with everything mentioned above to master branch

fleupold replied:
Why not to this PR branch? This allows me to review just the changes we discussed here and comment on them (now I need to find the relevant changes in a commit titled "A lot of QOL changes and leftover python implementation" which doesn't seem standard engineering practice)

rylaix replied:
My fault. Messed up with test branch difference.
Anyway, I will reset the branch now and create PR (so It would be easier to you:)

rylaix committed:
Python Implementation, SQL, and further Docker implementation.

Changelog:

Configurations:
Moved DUNE_QUERY_ID to a config.yaml
Added more configurations to config.yaml for further docker implementation.
Moved query ID's to config.yaml.

Included folder configurations where to store data and logs.
Included "validate_sql" to enable or disable our SQL's comparison.
Included "log_errors" to on or off logs.
Also, misc configurations related to multiprocessing (for data storing).

Queries:
Added Two SQL's discussed in PR (Also added a "fetch_latest_block.sql", so there will be from where to start).

Python:
Adapted "compare_and_validate_sql" func to use both of our queries.
Adapted "execute_query_and_get_results" to correctly use "start_block" and "end_block" params respectively (start_block is the latest gathered in data folder).

Split our Data gathering into two functions ("get_mev_blocker_bundles" to gather inside of a mevblocker, and "get_non_mev_transactions" to gather what is left.)

Mefiseru commented:
Added a commit.
Also, included Full changelog

fleupold commented on queries/fetch_latest_block.sql:
```diff
@@ -0,0 +1,14 @@
+SELECT 
```
Why is this query needed? Couldn't we simply use `u64::max` for `end_block` if we want all data until now? Also in terms of performance we probably need some kind of pagination mechanism (but I'm not sure).

fleupold commented on queries/fetch_remaining_transactions.sql:
```diff
@@ -0,0 +1,16 @@
+SELECT 
```
This query doesn't work (it has syntax errors).

Also, what is the data you need from the other confirmed transactions in the block for milestone 2?

fleupold commented on config/config.yaml:
```diff
@@ -1,4 +1,22 @@
-dune_query_id: 2446884
-polling_rate_minutes: 2 # Polling interval in minutes
+dune_query_id: 3993272  # The Dune query ID used for fetching data
+second_dune_query_id: 3993271
+gather_to_start_dune_query_id: 3992306
```
Unclear names. I think you only need two queries (and should name the parameters more expressively)
- All MEV Blocker bundle per block
- Confirmed transactions per block

Mefiseru commented:
Are we good to go with this query for fetching remaining transactions (for later simulating)? (Have we to adapt our mevblocker query to gather this fields to the same purpose too?)
```SQL
SELECT 
    e.block_number, 
    e.timestamp,
    e.hash,
    e.from,
    e.to,
    e.value,
    e.gas,
    e.gas_price
FROM 
    ethereum.transactions e
WHERE 
    e.block_number >= {{start_block}} 
    AND e.block_number <= {{end_block}} 
    AND e.hash NOT IN (
        SELECT DISTINCT from_hex(CAST(json_extract(transactions, '$[0].hash') AS VARCHAR)) 
        FROM mevblocker.raw_bundles t 
        WHERE t.blockNumber = e.block_number
    )
ORDER BY 
    e.block_number, e.hash;
```

fleupold commented:
Please take a look at the output of the MEVBlocker query you wrote for a sample range. You will see that all fields that are required for simulating are part of the json.

The query you outlined here is missing transaction index (to re-assemble the block ordering later) and calldata if I'm not mistaken. But again please take a look at the output and check if it has everything you need to simulate a counter factual block later on. 

Also I don't think you need to exclude the MEV blocker transactions here so that it's easy later on to decide where the bundle has to be "stitched in" in the simulation. 

Overall, I'm not sure you have clarity on what data is required to continue with the second milestone (we are really discussing things on a micro level, which shouldn't be necessary for such a grant). 

Can you maybe take a step back and take another look at the grant description to make sure it's clear what we are trying to build here and to ensure you can continue more autonomously?

rylaix committed:
Changed query names in config.yaml. Changed query.

Also, made small changes to the data_gathering.py itself (mainly related to the params).

Mefiseru commented:
You're absolutely right about the need for the transaction index and calldata. I've included both in fetch_remaining_transactions.sql  query to ensure we have all the necessary data for reassembling block ordering and for the counterfactual simulations required in Milestone 2.

Regarding the exclusion of MEV Blocker transactions, you make a good point. While we initially separated them out using fetch_backruns.sql, it might indeed be simpler and more efficient to gather everything in one query without excluding MEV Blocker transactions. This approach would reduce complexity in the data analysis and simulation stages.

As for my clarity on the data requirements, your feedback has been very helpful in ensuring that I’m covering all bases. I initially wanted to make sure I wasn't missing any crucial data points, and your input helped me identify the gaps I needed to address. With these adjustments, I believe we're now aligned with the requirements for Milestone 2.

c43aaa24e51da804df5a983d36a80f3b5ee34e8a done:)

rylaix committed:
Added "nonce" field to query.

Nonce: Adding the nonce field ensures that transactions are processed in the correct order for a given sender, which is critical for accurately simulating the transaction sequence.

Also, in future I'll probably will need to add fields related with "fee per gas"
Why? We're trying to Simulate scenarios where gas fee strategies play a critical role (e.g., optimizing the cost for users or understanding how miners choose transactions), then having max_priority_fee_per_gas along with max_fee_per_gas may provide a fuller picture.

Also for future considerations:
max_fee_per_gas: The maximum total amount (including base fee and priority fee) that a user is willing to pay per gas unit. This is also vital for simulating and evaluating transactions under EIP-1559.

base_fee_per_gas: Although this is determined by the network and not by individual transactions, it could be beneficial to fetch it to understand the relationship between the base fee and the max fees offered by users (not crucial, but can be considerable).

rylaix committed:
Changes related with Error handling and some unusual cases (see desc.)

Changelog:
Enhanced a logic related with "start_block_offset" and case in which there's no gathered block in "data"

Added a Logic which will pass to the second query If 1st returned 0 results.

Deleted "gather_to_start.py" as It's now not needed at all in all cases.
Deleted "fetch_latest_block.sql" as It's now not needed.

rylaix committed:
Adjusted Query ID's

fleupold commented on queries/fetch_remaining_transactions.sql:
```diff
@@ -0,0 +1,19 @@
+    e."from",
+    e."to",
+    e.value,
+    e.gas_used AS gas,
+    e.gas_price,
```
I think for simulation you need `gas_limit` and the EIP-1559 values for the gas price (for Type 2 transactions).

fleupold commented on src/data_gathering.py:
```diff
@@ -146,22 +227,38 @@ def process_block(block_number, bundles):
-    block_numbers = [latest_block_number - i for i in range(num_blocks_to_process)]  # Process the configured number of blocks
+    # Determine the number of blocks to process
+    latest_block_number = web3.eth.block_number
+    latest_processed_block = get_latest_processed_block() or latest_block_number - config.get('start_block_offset', 100)
+
+    # Check if `num_blocks_to_process` is set to "all"
+    num_blocks_to_process = config.get('num_blocks_to_process', 5)
+    if num_blocks_to_process == "all":
```
This logic is quite implicit and not really intuitive (fetch last 100 blocks in intervals of 5 when nothing is specified). Can we make sure we log the range for which the script is running so that this is easily observable in the output?

Also, I think it should be possible to specify a specific start block somehow (in case there has been a big gap we want to backfill).

maybe just having start block (with default last processed if found, otherwise current-block 100), end block (with default current block) and chunk size (default whatever is reasonable with a free dune account) makes sense?

fleupold reviewed (approved):
Query looks good (some small comments inline). If the second dune query is expensive credit wise, we could also gather this data from the node directly (as you did in the first delivery), but let's continue as is for now.

rylaix committed:
Merge branch 'main' into tests

Mefiseru commented:
@fleupold Proceeding with last commit in new RP, and Milestone 2?
